src/core/Akka.Streams/Implementation/SubscriberManagement.cs

```diff
@@ -313,10 +313,9 @@ protected void PushToDownstream(T value)
             if (_endOfStream is SubscriberManagement.NotReached)
             {
                 _pendingFromUpstream--;
-                var oldBufferLength = _buffer.Value.Length;
                 if (!_buffer.Value.Write(value))
                     throw new IllegalStateException("Output buffer overflow");
-                if (_buffer.Value.Length > oldBufferLength && Dispatch(_subscriptions))
+                if (_buffer.Value.Count() > 0 && Dispatch(_subscriptions))
```

Comment:
On a distinct value buffer, a `Write()` might succeed without adding any data into the buffer, so checking if a `Write()` succeed or not is not a guaranteed guard, we also need to need to check if the buffer actually have data to send.

Reply:
Does `_buffer` get modified by `Dispatch` ?

Reply:
Yes, `Dispatch()` calls `_buffer.Value.Read()` which modifies the buffer. The buffer will throw an exception if it doesn't have any data to read.

Reply:
Ok, so `_buffer.Value.Count() > 0` can never go from `false` -> `true` when `Dispatch` gets invoked?

Reply:
`Dispatch()` will never get called if `_buffer.Value.Count()` returns 0 because the boolean expression will get shortcutted?

Reply:
What I mean is - is it possible that `Dispatch` adds the item into the buffer under any circumstances? Just trying to head off any bugs that could cause a "stuck" Akka.Streams stage.

Reply:
Ah, no, `Dispatch()` only does a single read, it doesn't write to the buffer

Reply:
You answered my question in slack - `Dispatch` only performs a single read from the buffer. In this case, this should be fine.